server.py

```diff
@@ -386,6 +393,115 @@ async def create_cluster(  # pylint: disable=too-many-arguments,too-many-positio
+    yaml. Then you can tweak it as needed. Do not generate nmstate yaml from scratch without calling
+    this tool.
+
+    Returns: the generated nmstate yaml
+
+    Input param schema:
+    {NMStateTemplateParams.model_json_schema()}
+"""
+)
+@track_tool_usage()
+async def generate_nmstate_yaml(params: NMStateTemplateParams) -> str:
+    """
+    Generate an initial nmstate yaml.
+
+    See the mcp.tool description for more details (we have to use it since we dynamically generate
+    the input schema for the params).
+    """
+    log.info("Generate nmstate yaml with params: %s", params.model_dump_json(indent=2))
+    try:
+        generated = generate_nmstate_from_template(params)
+        log.debug("Generated yaml: %s", generated)
+        return generated
+    except TemplateError as e:
+        log.error("Failed to render nmstate template", exc_info=e)
+        return "ERROR: Failed to generate nmstate yaml"
+    except Exception as e:
+        log.error("Exception generating nmstate yaml", exc_info=e)
+        return "ERROR: Unknown error"
+
+
+@mcp.tool()
+@track_tool_usage()
+async def alter_static_network_config_nmstate_for_host(
+    cluster_id: str, index: int | None, new_nmstate_yaml: str | None
+) -> str:
+    """
+    Add, replace, or delete the nmstate yaml for a single host.
+
+    Args:
+        cluster_id (str): The unique identifier of the cluster
+        index (int | None): The index of the host in the existing static config to replace, or None to append a new host to the end of the config.
+        new_nmstate_yaml (str): The new nmstate YAML for a host. Leave this None to delete the config at the given index.
+
+    Returns: the updated infra env with the new static network config
+    """
+    client = InventoryClient(get_access_token())
+    infra_env_id = await _get_cluster_infra_env_id(client, cluster_id)
+    infra_env = await client.get_infra_env(infra_env_id)
+
+    if new_nmstate_yaml is None:
+        if index is None:
+            raise ValueError("index cannot be null when removing a host yaml")
+        if not infra_env.static_network_config:
+            raise ValueError(
+                "cannot remove host yaml with empty existing static network config"
+            )
+        static_network_config = remove_static_host_config_by_index(
+            existing_static_network_config=infra_env.static_network_config, index=index
+        )
+    else:
+        static_network_config = add_or_replace_static_host_config_yaml(
+            existing_static_network_config=infra_env.static_network_config,
+            index=index,
+            new_nmstate_yaml=new_nmstate_yaml,
+        )
+
+    result = await client.update_infra_env(
+        infra_env_id, static_network_config=static_network_config
+    )
+    return result.to_str()
+
+
+@mcp.tool()
+@track_tool_usage()
+async def list_infra_envs(cluster_id: str) -> str:
+    """
+    List all of the infra_envs associated with the given cluster_id.
```

Comment:
We've so far tried pretty hard to hide infraenv as a concept in this server. Should this maybe be getting the static network config for a cluster instead?